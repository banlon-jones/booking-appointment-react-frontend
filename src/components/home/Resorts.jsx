import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { useDispatch, useSelector } from 'react-redux';
import { GrNext, GrPrevious } from 'react-icons/gr';
import classes from './Resorts.module.css';

import { resortsFetched } from '../../store/resorts/resortsSlice';
import CircleSpinner from '../spinners/CircleSpinner';
import Resort from './Resort';

const Resorts = () => {
  const resorts = useSelector((state) => state.resorts);
  const dispatch = useDispatch();

  const [errorMessage, setErrorMessage] = useState('');
  const [loading, setLoading] = useState(false);
  const [currentPage, setCurrentPage] = useState(0);
  const [currentItems, setCurrentItems] = useState([]);
  const [pageSize, setPageSize] = useState(3);

  const baseIndex = () => currentPage * pageSize;

  const prevClickHandler = () => {
    if (currentPage > 0) setCurrentPage((currentPage) => currentPage - 1);
    else setCurrentPage(0);
  };

  const nextClickHandler = () => {
    if (currentPage < Math.ceil(resorts.length / pageSize) - 1) {
      setCurrentPage((currentPage) => currentPage + 1);
    }
  };

  const fetchResorts = async () => {
    try {
      setLoading(true);
      const { data } = await axios({
        url: 'https://resorts-booking-api.herokuapp.com/resorts',
        method: 'GET',
        headers: {
          authorization: sessionStorage.getItem('JwtAccessToken'),
        },
      });
      if (data.error) {
        setErrorMessage(data.error);
        return;
      }
      dispatch(resortsFetched(data));
    } catch (err) {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchResorts();
    if (resorts) {
      setLoading(false);
    }
    setPageSize(3);
  }, []);

  useEffect(() => {
    setCurrentItems(resorts.slice(baseIndex(), baseIndex() + pageSize));
  }, [resorts]);

  useEffect(() => {
    setCurrentItems(resorts.slice(baseIndex(), baseIndex() + pageSize));
  }, [currentPage]);

  return (
    <div className={classes.resorts_container}>
      <div className={classes.resorts_header}>
        <h1>Most Visited Resorts</h1>
        <p>Please select a resort to visit.</p>
      </div>
      <div className={classes.current_resorts_container}>
        <button
          type="button"
          onClick={() => prevClickHandler()}
          className={(currentPage > 0) ? classes.prev : classes.prev_disabled}
          disabled={(currentPage <= 0)}
        >
          <GrPrevious />
        </button>
        <ul className={classes.current_resorts_list}>
          {currentItems.map((resort) => (
            <li key={resort.id}>
              <Resort resort={resort} />
            </li>
          ))}
        </ul>
        <button
          type="button"
          onClick={() => nextClickHandler()}
          className={
            (currentPage < Math.ceil(resorts.length / pageSize) - 1)
              ? classes.next
              : classes.next_disabled
            }
          disabled={currentPage >= Math.ceil(resorts.length / pageSize) - 1}
        >
          <GrNext />
        </button>
      </div>
      {errorMessage && <h2>{errorMessage}</h2>}
      {loading && <CircleSpinner />}
    </div>
  );
};

export default Resorts;
