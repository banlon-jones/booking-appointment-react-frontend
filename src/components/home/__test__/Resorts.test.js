import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';

import Resorts from '../Resorts';

describe('Resorts snapshots', () => {
  const initialState = { resorts: [] };
  const mockStore = configureStore();
  let store;

  it('matches Resorts snapshot', () => {
    store = mockStore(initialState);
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Resorts />
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('Page rendering', () => {
    beforeEach(() => {
      render(
        <Provider store={store}>
          <Router>
            <Resorts />
          </Router>
        </Provider>,
      );
    });

    test('renders the Resorts page heading', () => {
      expect(screen.getByRole('heading')).toHaveTextContent(/Most Visited Resorts/);
    });

    test('renders button in the ResortsToDelete page', () => {
      expect(screen.getAllByRole('button')).toHaveLength(2);
    });
    test('renders unordered list item in the ResortsToDelete page', () => {
      expect(screen.getByRole('list'));
    });
  });
});
