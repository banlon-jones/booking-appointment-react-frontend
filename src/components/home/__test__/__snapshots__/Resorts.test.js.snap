// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resorts snapshots matches Resorts snapshot 1`] = `
<div
  className="resorts_container"
>
  <div
    className="resorts_header"
  >
    <h1>
      Most Visited Resorts
    </h1>
    <p>
      Please select a resort to visit.
    </p>
  </div>
  <div
    className="current_resorts_container"
  >
    <button
      className="prev_disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          fill="none"
          points="7 2 17 12 7 22"
          stroke="#000"
          strokeWidth="2"
          transform="matrix(-1 0 0 1 24 0)"
        />
      </svg>
    </button>
    <ul
      className="current_resorts_list"
    />
    <button
      className="next_disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          fill="none"
          points="7 2 17 12 7 22"
          stroke="#000"
          strokeWidth="2"
        />
      </svg>
    </button>
  </div>
</div>
`;
